source/shared/core_conn.cpp

```diff
@@ -150,34 +150,125 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
+            // In non-Windows environment, unixODBC 2.3.4 and unixODBC 2.3.1 return different error states 
+            // If it fails to connect using ODBC 17, it is unreliable to check for a certain sql state error.  
+            // Thus, we will simply throw an exception.
+            if( core_search_odbc_driver_unix( ODBC_DRIVER_17 ) ) {
+                conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ODBC_DRIVER_17];
+                r = core_odbc_connect( conn, conn_str, is_pooled );
+            }
+            else {
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_CE_DRIVER_REQUIRED, get_processor_arch()) {
+                    throw core::CoreException();
+                }
+            }
+        } // else driver_version not unknown
     }
-    else if ( conn->ce_option.enabled ) {
-        conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION::ODBC_DRIVER_17 ];
-        r = core_odbc_connect( conn, conn_str, is_missing_driver, is_pooled );
+    else { 
+        // column encryption NOT enabled
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
+            // check if the ODBC driver actually exists, if not, throw an exception
+            CHECK_CUSTOM_ERROR( ! core_search_odbc_driver_unix( conn->driver_version ), conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND ) {
+                throw core::CoreException();
+            }
 
-        CHECK_CUSTOM_ERROR( is_missing_driver, conn, SQLSRV_ERROR_AE_DRIVER_NOT_INSTALLED, get_processor_arch()) {
-            throw core::CoreException();
+            r = core_odbc_connect( conn, conn_str, is_pooled );
         }
-    }
-    else {
+        else {
+            // skip ODBC 11 in a non-Windows environment -- only available in Red Hat / SUSE (preview)
+            // https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server#microsoft-odbc-driver-11-for-sql-server-on-linux
+
+            DRIVER_VERSION odbc_version = ODBC_DRIVER_UNKNOWN;
+            if( core_search_odbc_driver_unix( ODBC_DRIVER_13 ) ) {
+                odbc_version = ODBC_DRIVER_13;
+            } 
+            else if ( core_search_odbc_driver_unix( ODBC_DRIVER_17 ) ) {
+                odbc_version = ODBC_DRIVER_17;
+            }
+
+            CHECK_CUSTOM_ERROR( odbc_version == ODBC_DRIVER_UNKNOWN, conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch() ) {
+                throw core::CoreException();
+            }
+            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[odbc_version];
+            r = core_odbc_connect( conn, conn_str_driver, is_pooled );
+        }  // else driver_version not unknown
+    } // else ce_option enabled
+#else
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+
+            if( core_compare_error_state( conn, r, "IM002" ) ) {
+                // sql state IM002 means that the specified ODBC driver is not installed
```

Comment:
IM002 means Data source not found and no default driver specified. You could get this if the driver was installed but the DSN specified isn't defined or the DSN specifies a path to the driver that's wrong.